In [1]:
### TODO:
# 1. Do feature engineering on group level instead of user level
# 2. Separate data between first person mode and free for all mode
# 3. Eliminate cheaters and anomalies
# 4. Develop prediction funcgion
#    - Final ranking per match can be determined using this formula -> 100/maxPlace, as ranking interval

In [4]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
# import xgboost
from matplotlib import pyplot as plt

%precision %.4f

'%.4f'

In [6]:
### Read training data
train = pd.read_csv('../../input/train_V2.csv')

In [8]:
train['winPlacePerc'].values

array([0.4444, 0.64  , 0.7755, ..., 0.4815, 0.8   , 0.5464])

In [ ]:
train.columns

In [ ]:
train.head(5)

In [ ]:
from pylab import rcParams
rcParams['figure.figsize'] = 10, 8

In [ ]:
matchTypes = train.groupby('matchId')['matchType'].first().value_counts()

In [ ]:
matchTypes.index

In [ ]:
fig, ax = plt.subplots()
ax.bar(np.arange(len(matchTypes.index)),matchTypes.values,align='center')
ax.set_xticks(np.arange(len(matchTypes.index)))
ax.set_xticklabels(matchTypes.index,rotation=70)
plt.title("Match type queue distribution")
plt.show()


### 1. Group Level Inspection and Feature Engineering

Features need to be generated on a group level. Below is the explanation. Let's inspect data for matchId **a10357fd1a4a91**

In [ ]:
### Inspect a match data
matchA_df = train[train.matchId == 'a10357fd1a4a91']

In [ ]:
### Inspect a group
matchA_df[matchA_df.groupId == '654c638629b8fc']

In [ ]:
### select groupIds in a match
playerGroups = matchA_df[['Id','groupId']]

In [ ]:
### Number of players per group
playerCountGroup = playerGroups.groupby('groupId',as_index=False).agg({'Id':'count'}).sort_values('Id').rename(columns={"Id":"players"}).reset_index(drop=True)

In [ ]:
playerCountGroup

Various number of players in a group may exist in one match. Player queuing solo will be in a group with only 1 member.

In [ ]:
### Total players in this match
print("Total players: {}".format(playerCountGroup.players.sum()))
print("Total groups: {}".format(playerCountGroup.groupId.count()))

Now notice the distribution of placement ranking which we would like to predict

In [ ]:
### Notice the percentage ranking
matchA_df[['winPlacePerc']].drop_duplicates().sort_values('winPlacePerc').reset_index(drop=True)

There are 26 placements which actually correspond to the number of groups. The interval could be retrieved using:

In [ ]:
### The increment of winPlacePerc is retrieved using:
print((100/float(26))/100)

Because the ranking is spread based on the number of groups in one match, group level features need to be generated!

In [ ]:
### Generate group level features    

def generate_group_level_features(dataset,feature_columns=['kills','assists','boosts']):
    features = dataset[["matchId","groupId",*feature_columns]].reset_index(drop=True)
    matchGroups = features[["matchId","groupId"]].drop_duplicates().reset_index(drop=True)
    
    ### predefined basic statistic operations
    _stats = ['max','min','sum','mean','std']
    
    ### calculate group level features
    for f in feature_columns:
        for s in _stats:
            new_field = '{s}_{f}'.format(s=s,f=f)
            print(new_field)
            matchGroups = pd.merge(matchGroups,
                features.groupby(["matchId","groupId"],as_index=False)\
                .agg({f:s}).rename(columns={f:new_field}).fillna(0)[["matchId","groupId",new_field]].drop_duplicates(),
                on=['matchId','groupId'],how='inner'
            )
            
    return matchGroups.reset_index(drop=True)

In [ ]:
### sample of group level features
#generate_group_level_features(matchA_df)

In [ ]:
import time
s = time.time()
groupLevelFeatures_train = generate_group_level_features(train)
e = time.time()
print("elapsed {}s".format(e-s))

In [ ]:
groupLevelFeatures_train.to_csv("groupLevelFeatures_train.csv",index=False)

In [ ]:
# dummy
# a = pd.DataFrame(data=[{"a":1,"b":2},{"a":3,"b":6}])

# for i,r in a.iterrows():
#     print(r['a'])

# def dum(x):
#     x['c'] = x['a'] + x['b']
#     return x

# a = a.apply(lambda x: dum(x),1)

# a.iloc[0]['c'] = 12

# a

### 2. Separate Game Modes data

(TODO) <br>
There are several game modes / match types in PUBG <br>
https://pubg.gamepedia.com/Game_Modes <br>
Patterns might differ for example between First Person Mode and Third Person Mode even though the players are on solo queue game.

### 3. Eliminate Anomalies

(TODO) <br>
There are already existing kernel out there mentioning anomalies or cheaters in PUBG matches. <br>
We need to adopt some of them.

### 4. Prediction Functions

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, f1_score, mean_absolute_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

In [4]:
# groupLevelFeatures_train.head()

In [5]:
# len(groupLevelFeatures_train['matchId'].unique()), len(groupLevelFeatures_train['groupId'].unique())

In [6]:
# train.head()
train.sort_values(['matchId', 'winPlacePerc']).head()

Id         groupId         matchId  assists  boosts  \
3599150  fc62a751955351  3a6addfa0df938  0000a43bce5eec        0       0   
3895898  7ad5883d71d42e  3a6addfa0df938  0000a43bce5eec        0       0   
1425950  5a3afae17b53c0  236ab9e9c081b9  0000a43bce5eec        0       0   
1998699  e431d8a8f6c99b  236ab9e9c081b9  0000a43bce5eec        0       0   
2290792  e7b325d63d8393  236ab9e9c081b9  0000a43bce5eec        0       0   

         damageDealt  DBNOs  headshotKills  heals  killPlace      ...       \
3599150         0.00      0              0      0         95      ...        
3895898         0.00      0              0      0         94      ...        
1425950        66.65      0              0      0         93      ...        
1998699        21.29      0              0      0         92      ...        
2290792         0.00      0              0      0         91      ...        

         revives  rideDistance  roadKills  swimDistance  teamKills  \
3599150        0           0.0          0           0.0          0   
3895898        0           0.0          0           0.0          0   
1425950        0           0.0          0           0.0          0   
1998699        0           0.0          0           0.0          0   
2290792        0           0.0          0           0.0          0   

        vehicleDestroys  walkDistance  weaponsAcquired  winPoints  \
3599150               0         9.636                1       1557   
3895898               0         0.000                0       1413   
1425950               0        17.510                1       1430   
1998699               0         5.201                1       1503   
2290792               0        18.650                1       1450   

         winPlacePerc  
3599150         0.000  
3895898         0.000  
1425950         0.037  
1998699         0.037  
2290792         0.037  

[5 rows x 29 columns]

In [7]:
train = train[train['winPlacePerc'].notnull()].reset_index(drop=True)

In [66]:
# col_metrics = ['assists', 'boosts', 'damageDealt', 'DBNOs',
#        'headshotKills', 'heals', 'killPlace', 'killPoints', 'kills',
#        'killStreaks', 'longestKill', 'matchDuration', 'maxPlace',
#        'numGroups', 'rankPoints', 'revives', 'rideDistance', 'roadKills',
#        'swimDistance', 'teamKills', 'vehicleDestroys', 'walkDistance',
#        'weaponsAcquired', 'winPoints']
# col_drop = ['Id', 'groupId', 'matchId']

col_metrics = ['boosts', 'damageDealt', 'DBNOs',
       'headshotKills', 'killPlace', 
       'killStreaks', 'longestKill',
       'walkDistance',
       'weaponsAcquired']
col_drop = ['Id', 'groupId', 'matchId']

In [67]:
X_train, X_test, y_train, y_test = train_test_split(train.drop('winPlacePerc', axis=1), 
                                    train[['Id', 'winPlacePerc']], test_size=0.33, random_state=42)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((2979466, 28), (1467499, 28), (2979466, 2), (1467499, 2))

In [68]:
X_train_2 = X_train.drop(col_drop, axis=1)
X_train_2.head()

assists  boosts  damageDealt  DBNOs  headshotKills  heals  killPlace  \
1486370        0       1        449.6      5              1      1          6   
1208781        1       0         87.9      0              0      0         74   
3427006        0       0        180.5      0              0      0         36   
1658192        0       0        387.2      3              1      0         13   
2555464        0       0          0.0      0              0      0         79   

         killPoints  kills  killStreaks    ...      rankPoints  revives  \
1486370           0      4            1    ...            1500        0   
1208781           0      0            0    ...            1445        0   
3427006           0      1            1    ...            1504        0   
1658192        1547      3            2    ...              -1        0   
2555464           0      0            0    ...            1506        0   

        rideDistance  roadKills  swimDistance  teamKills  vehicleDestroys  \
1486370          0.0          0           0.0          0                0   
1208781        151.8          0           0.0          0                0   
3427006          0.0          0           0.0          0                0   
1658192          0.0          0           0.0          0                0   
2555464          0.0          0           0.0          0                0   

         walkDistance  weaponsAcquired  winPoints  
1486370        1455.0                5          0  
1208781         379.9                3          0  
3427006         298.5                2          0  
1658192         129.0                1       1562  
2555464         120.9                3          0  

[5 rows x 25 columns]

# Decision Tree

In [69]:
classifier = DecisionTreeRegressor(random_state=42)
classifier.fit(X_train_2[col_metrics], y_train['winPlacePerc'])
del X_train_2

# Random Forest

In [70]:
# classifier = RandomForestRegressor(n_jobs=4, n_estimators=10, random_state=42)
# classifier.fit(X_train_2[col_metrics], y_train['winPlacePerc'])
# del X_train_2

In [97]:
X_test_2 = X_test.drop(col_drop, axis=1)
y_pred = classifier.predict(X_test_2[col_metrics])
del X_test_2
y_pred, X_test.shape, len(y_pred)

(array([0.5208, 0.9053, 0.7938, ..., 0.8125, 0.1481, 0.9381]),
 (1467499, 28),
 1467499)

In [72]:
mean_absolute_error(y_test['winPlacePerc'], y_pred)

0.10329007607145285

In [73]:
X_test_copy = X_test.copy()
X_test_grp = X_test[['matchId','groupId']].copy()
X_test_copy.drop(['matchId','groupId'], axis=1, inplace=True)
X_test_grp['winPlacePerc'] = y_pred
X_test_grp['y_pred'] = y_pred
X_test_grp.head()

matchId         groupId  winPlacePerc  y_pred
4270779  e1c02c1421bdd4  6dfb23f52dc772        0.5208  0.5208
1687951  7669a0409fa32a  43b509a591d8f1        0.9053  0.9053
3732582  a3eae2eaede76e  3d2944992d2111        0.7938  0.7938
3386216  a22b2e7366205a  405f55ab35614a        0.1099  0.1099
243580   7f9e3909d86e2e  cb396c5f1bb96e        0.7609  0.7609

In [74]:
# X_test_copy.head()
y_test['winPlacePerc_true'] = y_test['winPlacePerc']
y_test.drop(['winPlacePerc'], axis=1, inplace=True)
y_test.head()

Id  winPlacePerc_true
4270779  ef2d51d18a38cf             0.2800
1687951  0ccf6e163b5009             0.8571
3732582  948c2d231060b2             0.7812
3386216  3a5872249139fc             0.1379
243580   a78cc239213b35             0.7263

In [75]:
group = X_test_grp.groupby('matchId')
X_test_grp['_rank.winPlacePerc'] = group['winPlacePerc'].rank(method='min')
print(X_test_grp[['matchId', 'groupId', 
                  '_rank.winPlacePerc', 'y_pred']].sort_values(['matchId', '_rank.winPlacePerc']).head())
X_test_copy = pd.concat([X_test_copy, X_test_grp], axis=1)
X_test_copy = pd.merge(X_test_copy, y_test)
X_test_copy.sort_values('Id').head()

                matchId         groupId  _rank.winPlacePerc    y_pred
609078   0000a43bce5eec  e8ff1c0fe7f6aa                 1.0  0.000000
3599149  0000a43bce5eec  3a6addfa0df938                 1.0  0.000000
1998699  0000a43bce5eec  236ab9e9c081b9                 1.0  0.000000
3895897  0000a43bce5eec  3a6addfa0df938                 4.0  0.002988
4367295  0000a43bce5eec  236ab9e9c081b9                 5.0  0.044400


Id  assists  boosts  damageDealt  DBNOs  headshotKills  \
838752   0000053e9d70a9        0       2        183.0      0              1   
1056895  00002299115341        0       1        114.7      0              0   
747646   000035d735aafb        0       1        100.0      1              0   
1444670  000044511e367a        1       2        224.6      3              1   
149030   000056883d8510        0       0          0.0      0              0   

         heals  killPlace  killPoints  kills        ...          \
838752       1         15           0      2        ...           
1056895      0         86           0      0        ...           
747646       0         74           0      0        ...           
1444670      1          9           0      3        ...           
149030       0         70           0      0        ...           

         vehicleDestroys  walkDistance  weaponsAcquired winPoints  \
838752                 0       2065.00                2         0   
1056895                0        125.10                2         0   
747646                 0        432.10                6         0   
1444670                0       2757.00                3         0   
149030                 0         31.24                1         0   

                matchId         groupId  winPlacePerc  y_pred  \
838752   e65e89709875ca  ea6619b280f5c9        0.9684  0.9684   
1056895  67402a2f879c8f  92041a706ac1ff        0.1667  0.1667   
747646   f15961b94c9268  4b474ef3331a92        0.1786  0.1786   
1444670  e946a37ec1dc3f  61562e23b34da7        0.9200  0.9200   
149030   ca72d76111cf62  8d0fe23814f136        0.2692  0.2692   

         _rank.winPlacePerc  winPlacePerc_true  
838752                 31.0             0.8125  
1056895                 2.0             0.0000  
747646                  5.0             0.2222  
1444670                24.0             0.8889  
149030                 13.0             0.3333  

[5 rows x 32 columns]

In [76]:
# X_test_copy[['Id', 'numGroups', 'maxPlace', 'winPlacePerc', '_rank.winPlacePerc']].head(10)

In [77]:
fullgroup = (X_test_copy['numGroups'] == X_test_copy['maxPlace'])

print(sum(fullgroup))
# full group (201366) --> calculate from rank
subset = X_test_copy.loc[fullgroup]
X_test_copy.loc[fullgroup, 'winPlacePerc'] = (subset['_rank.winPlacePerc'].values - 1) / (subset['maxPlace'].values - 1)

# not full group (684872) --> align with maxPlace
subset = X_test_copy.loc[~fullgroup]
gap = 1.0 / (subset['maxPlace'].values - 1)
new_perc = np.around(subset['winPlacePerc'].values / gap) * gap  # half&up
X_test_copy.loc[~fullgroup, 'winPlacePerc'] = new_perc

X_test_copy['winPlacePerc'] = X_test_copy['winPlacePerc'].clip(lower=0,upper=1)


408636


In [96]:
print(X_test_copy.loc[~fullgroup]['matchId'].unique())
X_test_copy[['Id', 'matchId', 'groupId', 'winPlacePerc', 'y_pred', 
                   '_rank.winPlacePerc', 'winPlacePerc_true', '_pred.winPlace']].sort_values(['matchId', 'groupId']).head(50)

['7669a0409fa32a' 'a3eae2eaede76e' 'a22b2e7366205a' ... 'ba580ea8f2f61b'
 '5ddd684bf067f8' 'd6be8e6354f722']


Id         matchId         groupId  winPlacePerc  \
795458   1845bd80ef1e31  0000a43bce5eec  236ab9e9c081b9      0.148148   
1268099  e431d8a8f6c99b  0000a43bce5eec  236ab9e9c081b9      0.000000   
513634   7ad5883d71d42e  0000a43bce5eec  3a6addfa0df938      0.111111   
1095864  fc62a751955351  0000a43bce5eec  3a6addfa0df938      0.000000   
996011   9db8cb6643530a  0000a43bce5eec  4d1bbbc19b9084      1.000000   
1318594  ffc5bf94121ebc  0000a43bce5eec  599d924f8a02db      0.925926   
3051     a638435c730f4e  0000a43bce5eec  6620b219ed2ee2      0.814815   
387568   93825c1a755f53  0000a43bce5eec  6620b219ed2ee2      0.777778   
447386   51bf1d40b394c6  0000a43bce5eec  6c44ef4381fe8d      0.555556   
1235761  8c5828852148b5  0000a43bce5eec  767819928e6279      0.481481   
281826   02035450b9c95a  0000a43bce5eec  7bd08592bb25e2      0.962963   
935003   a86e30ee9cc26f  0000a43bce5eec  7bd08592bb25e2      0.444444   
614115   f99da346f1a999  0000a43bce5eec  8a2bf1adcd967d      0.740741   
1177349  3444098b1a9046  0000a43bce5eec  8ebb76fe544c27      0.851852   
310221   d73753fe20fa46  0000a43bce5eec  909fe185a8cdec      0.333333   
331611   f5827d148c67b4  0000a43bce5eec  909fe185a8cdec      0.407407   
691182   0d3fbb7d4b168c  0000a43bce5eec  909fe185a8cdec      0.259259   
1363087  f806332c66c741  0000a43bce5eec  909fe185a8cdec      0.592593   
47926    997b90d5130755  0000a43bce5eec  97364d4199bf30      0.185185   
361090   2e13f3a9112980  0000a43bce5eec  9cf70fed874fa1      0.222222   
1014550  de3ddbf280477e  0000a43bce5eec  9cf70fed874fa1      0.481481   
1184548  d38d809a2f91d7  0000a43bce5eec  9cf70fed874fa1      0.666667   
523612   7421fa7954935e  0000a43bce5eec  a572193b2af137      1.000000   
892272   737a4510a0a5ea  0000a43bce5eec  b431ffe23c623b      0.703704   
1184313  8eb803fb4b5d0a  0000a43bce5eec  b947dfc5cf2d02      0.888889   
961797   33f14597c50974  0000a43bce5eec  bf80d0d49c12dc      1.000000   
518929   770fdf888c099b  0000a43bce5eec  c0f2e306380d3c      1.000000   
779109   cc54bf99048d26  0000a43bce5eec  c0f2e306380d3c      1.000000   
1265911  dac594a2ca8fc5  0000a43bce5eec  c0f2e306380d3c      1.000000   
915733   58966681b2604c  0000a43bce5eec  ca058240046b4c      1.000000   
1152004  34a4c3d0305168  0000a43bce5eec  cf4d80e34833c2      0.629630   
1176434  857c375631842d  0000a43bce5eec  dc8f83946f97a9      0.370370   
1106531  51e37937b388fd  0000a43bce5eec  e84bd83822d783      0.259259   
991021   1b8c21ba00042e  0000a43bce5eec  e8ff1c0fe7f6aa      0.000000   
570224   30ec3d23b3045f  0000eb01ea6cdd  07192df36deae5      0.481481   
89096    ca77d9ef2b5e50  0000eb01ea6cdd  17e0f06615ab71      0.148148   
1093876  4463a3523b6d53  0000eb01ea6cdd  17e0f06615ab71      0.074074   
796641   a3a047de12edb6  0000eb01ea6cdd  1bedc8e1ee310b      0.185185   
1413317  26edfe5379c4bb  0000eb01ea6cdd  1bedc8e1ee310b      0.333333   
903981   260c550107cb06  0000eb01ea6cdd  1e304e14eb56b0      0.185185   
809202   a4228e2db2e7af  0000eb01ea6cdd  2dca719fa00654      0.962963   
1183445  075c5753f21496  0000eb01ea6cdd  338a69335655a6      0.629630   
286454   e2b1c128603900  0000eb01ea6cdd  6bea72cbdc149c      1.000000   
509246   477ec130f0dcb3  0000eb01ea6cdd  6bea72cbdc149c      0.814815   
66070    de0d69a7bfc693  0000eb01ea6cdd  8824c9f7406aca      0.000000   
84785    3064c4659e4d2a  0000eb01ea6cdd  8824c9f7406aca      0.000000   
1387117  6968f86c2974ed  0000eb01ea6cdd  8ba14e50fb3512      0.481481   
1104281  59999f7cabef78  0000eb01ea6cdd  9441c71b86b1fe      0.629630   
1106708  65691b9f9ee38c  0000eb01ea6cdd  ab2314d43d70dc      0.407407   
95546    f6e1aaed097f61  0000eb01ea6cdd  cd7d38d8566d58      0.666667   

           y_pred  _rank.winPlacePerc  winPlacePerc_true  _pred.winPlace  
795458   0.044400                 5.0             0.0370             NaN  
1268099  0.000000                 1.0             0.0370             NaN  
513634   0.002988                 4.0             0.0000             NaN

In [79]:
X_test_copy.loc[~fullgroup, '_pred.winPlace'] = np.around(X_test_copy.loc[~fullgroup, 'winPlacePerc'].values / gap) + 1
X_test_copy.loc[~fullgroup & (X_test_copy['matchId'] == '12acd71ccf720e'),
           ['matchId','groupId','winPlacePerc','maxPlace','numGroups','_pred.winPlace','_rank.winPlacePerc']
          ].sort_values(['matchId','_pred.winPlace','_rank.winPlacePerc'])

matchId         groupId  winPlacePerc  maxPlace  numGroups  \
459159   12acd71ccf720e  ab0ecc017f4a78      0.000000        96         92   
7        12acd71ccf720e  47458e09570b3f      0.000000        96         92   
430418   12acd71ccf720e  47458e09570b3f      0.021053        96         92   
476844   12acd71ccf720e  d1174936d621cd      0.021053        96         92   
657120   12acd71ccf720e  24666a091c34f1      0.052632        96         92   
1029598  12acd71ccf720e  f6a7f887aa3d95      0.073684        96         92   
1150802  12acd71ccf720e  d34c177da570eb      0.178947        96         92   
374318   12acd71ccf720e  d7ce9908c04deb      0.210526        96         92   
214278   12acd71ccf720e  260751ef30f92a      0.221053        96         92   
1084024  12acd71ccf720e  b96724d88e1285      0.242105        96         92   
1410854  12acd71ccf720e  ed6b877c9aa6e9      0.273684        96         92   
1306434  12acd71ccf720e  df554cb1949fb2      0.294737        96         92   
1303245  12acd71ccf720e  3ae93d2eb1eb84      0.357895        96         92   
977281   12acd71ccf720e  2b533694df61ee      0.389474        96         92   
1326881  12acd71ccf720e  ed426004163421      0.421053        96         92   
682825   12acd71ccf720e  a2ec6832317447      0.452632        96         92   
530257   12acd71ccf720e  ae6768919c38bb      0.484211        96         92   
1428397  12acd71ccf720e  a9ab12d7d23ff6      0.484211        96         92   
668116   12acd71ccf720e  dc4160acff070c      0.505263        96         92   
594671   12acd71ccf720e  bc7f45b5500f74      0.621053        96         92   
36653    12acd71ccf720e  d01c2cfa9c0f91      0.705263        96         92   
1223168  12acd71ccf720e  641b2bfd2d4696      0.726316        96         92   
71974    12acd71ccf720e  bae872d6515077      0.757895        96         92   
1069011  12acd71ccf720e  29b97a90283b17      0.757895        96         92   
644414   12acd71ccf720e  d87972c03b0f3f      0.800000        96         92   
534408   12acd71ccf720e  da9c3834a5f8d3      0.810526        96         92   
391483   12acd71ccf720e  b4c1e76c920ebc      0.873684        96         92   
941250   12acd71ccf720e  a243814fbac2e8      0.905263        96         92   
236073   12acd71ccf720e  8fc1f99428bc13      0.936842        96         92   
770196   12acd71ccf720e  b3be129bc14df4      0.947368        96         92   
82238    12acd71ccf720e  28ba2f3e3c2bea      0.968421        96         92   

         _pred.winPlace  _rank.winPlacePerc  
459159              1.0                 1.0  
7                   1.0                 2.0  
430418              3.0                 3.0  
476844              3.0                 4.0  
657120              6.0                 5.0  
1029598             8.0                 6.0  
1150802            18.0                 7.0  
374318             21.0                 8.0  
214278             22.0                 9.0  
1084024            24.0                10.0  
1410854            27.0                11.0  
1306434            29.0                12.0  
1303245            35.0                13.0  
977281             38.0                14.0  
1326881            41.0                15.0  
682825             44.0                16.0  
530257             47.0                17.0  
1428397            47.0                18.0  
668116             49.0                19.0  
594671             60.0                20.0  
36653              68.0                21.0  
1223168            70.0                22.0  
71974              73.0                23.0  
1069011            73.0                24.0  
644414             77.0                25.0  
534408             78.0                26.0  
391483             84.0                27.0  
941250             87.0                28.0  
236073             90.0                29.0  
770196             91.0                30.0  
82238              93.0                31.0

In [80]:
y_test.head(), X_test_copy['winPlacePerc'].head()

(                     Id  winPlacePerc_true
 4270779  ef2d51d18a38cf             0.2800
 1687951  0ccf6e163b5009             0.8571
 3732582  948c2d231060b2             0.7812
 3386216  3a5872249139fc             0.1379
 243580   a78cc239213b35             0.7263, 0    0.440000
 1    0.908163
 2    0.791667
 3    0.103448
 4    0.273684
 Name: winPlacePerc, dtype: float64)

In [84]:
final_df = pd.merge(y_test, X_test_copy[['Id', 'matchId', 'groupId', 'winPlacePerc', '_pred.winPlace', '_rank.winPlacePerc', 'y_pred']])
final_df.head(50)

Id  winPlacePerc_true         matchId         groupId  \
0   ef2d51d18a38cf             0.2800  e1c02c1421bdd4  6dfb23f52dc772   
1   0ccf6e163b5009             0.8571  7669a0409fa32a  43b509a591d8f1   
2   948c2d231060b2             0.7812  a3eae2eaede76e  3d2944992d2111   
3   3a5872249139fc             0.1379  a22b2e7366205a  405f55ab35614a   
4   a78cc239213b35             0.7263  7f9e3909d86e2e  cb396c5f1bb96e   
5   c9c0eddbf63319             0.1923  625cc032ef1253  9a78ed7d69a464   
6   9c754cf2e39552             0.9091  549073ce7cb13a  73b690d428df6e   
7   6f0a3fc5134110             0.0000  12acd71ccf720e  47458e09570b3f   
8   4f4a1fe62d1638             0.0833  a4f184e45b529c  2d5fcfc06e5a99   
9   660a11813c1435             0.1364  32b7d4d1461614  66fc34e84c2e09   
10  b0d4c2da1833e7             0.3214  1bba8976826d48  b9704fa3759c41   
11  aa87602a89b260             0.3793  5255c0f3de73a6  ae116c4e924ba3   
12  db83247a4ac3fc             0.5200  9d52f40fd47fa3  7ef4c030b6a14e   
13  e587faa90f33ab             0.0889  961f0789f13514  a93fbdd01227b1   
14  7fa0d3c38933f1             0.5510  71e6e082cafd61  63cad8470c8de2   
15  d3143943d62a78             0.2759  401a75ada5445b  d5821154b8f84e   
16  10e0e40494ff1d             0.8571  c532a600c7df08  4320599b9ea99a   
17  adb2a8d8e56323             0.8148  6e13e14b7bb499  84c60dbd11febc   
18  a822ffd64c69f2             0.6809  bd0fec60fbd344  a9cf4f082d22c8   
19  7b46f925998c02             0.4375  52bbca4abcd444  83fc5e77142896   
20  d8f7ed9d94cba0             0.4167  5a457c71251a3b  baee6fe4e3b9e5   
21  751b266f1f3c0c             0.6296  d9d9ac76fa5740  14fa4b0bd6ea02   
22  0726f741b4153a             0.8936  fb890442a17956  f6cd9563903ef8   
23  e8a0b6cda4f3c7             0.9247  470ede8ee5e357  c2907def772b22   
24  5ae614f6d15004             0.0690  c3a6dc85056330  24e6d677071606   
25  437f38fa676856             0.3804  ca1f833f0104ac  dd11f4e18ab7b0   
26  155bee6d75c10c             0.5957  a0cd013b4a9b92  0d5305a716ac0f   
27  aa1f068dd1d9ed             0.6531  d76536a02f0a4f  aaec777b5a52ee   
28  4b9d9eed859d01             0.7778  83542e5c7bf154  618be56b3a663e   
29  be7919bc31656b             0.5385  8995ed16853870  27cdff0bb9a9c5   
30  a14908331466f8             0.5670  d74a03eb59ba9e  b1660d1a6fab1d   
31  6529cca07557dd             0.4255  a1666c8d06f857  1dbae764d13fa1   
32  aeeeec9e6ee197             0.2391  46b58603a8e030  69b62248ae6f52   
33  dc7b892174cd20             0.3407  efe2b42795f8e9  c11f5e426207dd   
34  8a1c9a961898f8             0.1071  e66afe0a893eae  6bc5c58a843d0c   
35  163b28aeee44d6             0.4400  1c0629c79d0ceb  f71b55ffd6aadd   
36  030ff78474e77b             0.6667  d8ac0d11203bc9  ceb448ae848177   
37  443d469e9ad181             0.2759  460343f056e0a2  8fc104327b0cab   
38  e48b70391210d3             0.8737  71bc0fa4eccf17  08f51c0c90b039   
39  3d85980164e5cb             0.2800  f8a81be3ca8baa  4fe8859b0f36ab   
40  4a7f3c8514d66a             0.5926  4f79b0e12f7b96  36447766842cd3   
41  075722523d3f76             0.6429  77d2a5d2d81fe4  5b1d6b9bbbf393   
42  fa6bd484b37ae7             0.1735  38179d6de5e612  1859d51bbe19c9   
43  26d99a7613ebe1             0.0222  d4016baf55b46f  514f41f6b4b780   
44  b53a052943655c             0.3556  fc79d44590fe98  032b49a1d72355   
45  6f513afbc9bc0e             0.0968  c0912bab49544b  10ac98326019de   
46  55e417c779c1a4             0.6800  bf66b644a456d2  24a4c0f24281cd   
47  ab291ca890e6df             0.5833  a8f2faa993081f  d6d935afc77ece   
48  1eeb7edb6681b7             0.6667  40ef5bd1ea0375  c1051c3e2c15fe   
49  ac8d1921fdd163             0.1368  439ca4b590b371  14696695954b48   

    winPlacePerc  _pred.winPlace  _rank.winPlacePerc    y_pred  
0       0.440000             NaN                12.0  0.520800  
1       0.908163            90.0                26.0  0.905300  
2       0.791667            77.0                29.0  0.793800  
3       0.103448             4.0

In [87]:
final_df.sort_values(['matchId', 'groupId', 'winPlacePerc']).head(50)

Id  winPlacePerc_true         matchId         groupId  \
1268099  e431d8a8f6c99b             0.0370  0000a43bce5eec  236ab9e9c081b9   
795458   1845bd80ef1e31             0.0370  0000a43bce5eec  236ab9e9c081b9   
1095864  fc62a751955351             0.0000  0000a43bce5eec  3a6addfa0df938   
513634   7ad5883d71d42e             0.0000  0000a43bce5eec  3a6addfa0df938   
996011   9db8cb6643530a             1.0000  0000a43bce5eec  4d1bbbc19b9084   
1318594  ffc5bf94121ebc             0.5926  0000a43bce5eec  599d924f8a02db   
387568   93825c1a755f53             0.7778  0000a43bce5eec  6620b219ed2ee2   
3051     a638435c730f4e             0.7778  0000a43bce5eec  6620b219ed2ee2   
447386   51bf1d40b394c6             0.7037  0000a43bce5eec  6c44ef4381fe8d   
1235761  8c5828852148b5             0.2593  0000a43bce5eec  767819928e6279   
935003   a86e30ee9cc26f             0.6667  0000a43bce5eec  7bd08592bb25e2   
281826   02035450b9c95a             0.6667  0000a43bce5eec  7bd08592bb25e2   
614115   f99da346f1a999             0.5185  0000a43bce5eec  8a2bf1adcd967d   
1177349  3444098b1a9046             0.6296  0000a43bce5eec  8ebb76fe544c27   
691182   0d3fbb7d4b168c             0.1852  0000a43bce5eec  909fe185a8cdec   
310221   d73753fe20fa46             0.1852  0000a43bce5eec  909fe185a8cdec   
331611   f5827d148c67b4             0.1852  0000a43bce5eec  909fe185a8cdec   
1363087  f806332c66c741             0.1852  0000a43bce5eec  909fe185a8cdec   
47926    997b90d5130755             0.1111  0000a43bce5eec  97364d4199bf30   
361090   2e13f3a9112980             0.2963  0000a43bce5eec  9cf70fed874fa1   
1014550  de3ddbf280477e             0.2963  0000a43bce5eec  9cf70fed874fa1   
1184548  d38d809a2f91d7             0.2963  0000a43bce5eec  9cf70fed874fa1   
523612   7421fa7954935e             0.8519  0000a43bce5eec  a572193b2af137   
892272   737a4510a0a5ea             0.4074  0000a43bce5eec  b431ffe23c623b   
1184313  8eb803fb4b5d0a             0.7407  0000a43bce5eec  b947dfc5cf2d02   
961797   33f14597c50974             0.8148  0000a43bce5eec  bf80d0d49c12dc   
518929   770fdf888c099b             0.8889  0000a43bce5eec  c0f2e306380d3c   
779109   cc54bf99048d26             0.8889  0000a43bce5eec  c0f2e306380d3c   
1265911  dac594a2ca8fc5             0.8889  0000a43bce5eec  c0f2e306380d3c   
915733   58966681b2604c             0.9630  0000a43bce5eec  ca058240046b4c   
1152004  34a4c3d0305168             0.4444  0000a43bce5eec  cf4d80e34833c2   
1176434  857c375631842d             0.1481  0000a43bce5eec  dc8f83946f97a9   
1106531  51e37937b388fd             0.2222  0000a43bce5eec  e84bd83822d783   
991021   1b8c21ba00042e             0.0741  0000a43bce5eec  e8ff1c0fe7f6aa   
570224   30ec3d23b3045f             0.2963  0000eb01ea6cdd  07192df36deae5   
1093876  4463a3523b6d53             0.0741  0000eb01ea6cdd  17e0f06615ab71   
89096    ca77d9ef2b5e50             0.0741  0000eb01ea6cdd  17e0f06615ab71   
796641   a3a047de12edb6             0.3333  0000eb01ea6cdd  1bedc8e1ee310b   
1413317  26edfe5379c4bb             0.3333  0000eb01ea6cdd  1bedc8e1ee310b   
903981   260c550107cb06             0.2222  0000eb01ea6cdd  1e304e14eb56b0   
809202   a4228e2db2e7af             0.9259  0000eb01ea6cdd  2dca719fa00654   
1183445  075c5753f21496             0.5926  0000eb01ea6cdd  338a69335655a6   
509246   477ec130f0dcb3             0.9630  0000eb01ea6cdd  6bea72cbdc149c   
286454   e2b1c128603900             0.9630  0000eb01ea6cdd  6bea72cbdc149c   
66070    de0d69a7bfc693             0.0000  0000eb01ea6cdd  8824c9f7406aca   
84785    3064c4659e4d2a             0.0000  0000eb01ea6cdd  8824c9f7406aca   
1387117  6968f86c2974ed             0.5556  0000eb01ea6cdd  8ba14e50fb3512   
1104281  59999f7cabef78             0.7037  0000eb01ea6cdd  9441c71b86b1fe   
1106708  65691b9f9ee38c             0.2593  0000eb01ea6cdd  ab2314d43d70dc   
95546    f6e1aaed097f61             0.8148  0000eb01ea6cdd  cd7d38d8566d58   

         winPlacePerc  _pred.winPlace  _ra

In [90]:
X_test_copy[X_test_copy['groupId'] == '3a6addfa0df938']

Id  assists  boosts  damageDealt  DBNOs  headshotKills  \
513634   7ad5883d71d42e        0       0          0.0      0              0   
1095864  fc62a751955351        0       0          0.0      0              0   

         heals  killPlace  killPoints  kills       ...        walkDistance  \
513634       0         94        1306      0       ...               0.000   
1095864      0         95        1458      0       ...               9.636   

         weaponsAcquired  winPoints         matchId         groupId  \
513634                 0       1413  0000a43bce5eec  3a6addfa0df938   
1095864                1       1557  0000a43bce5eec  3a6addfa0df938   

         winPlacePerc    y_pred  _rank.winPlacePerc  winPlacePerc_true  \
513634       0.111111  0.002988                 4.0                0.0   
1095864      0.000000  0.000000                 1.0                0.0   

         _pred.winPlace  
513634              NaN  
1095864             NaN  

[2 rows x 33 columns]

In [83]:
mean_absolute_error(final_df['winPlacePerc'], final_df['winPlacePerc_true'])

0.11922039296180897

# Result

Attributes | MAE
--- | ---
**Decision Tree** |
All | 0.0822435372194249
All with Ranking | 0.020896060576532312, 0.10
All with Ranking (with ascending False) | 0.19069269005752446
Features from Chicken Dinner Notebook (including maxplace, numgroups) | 0.024114451660075895 
Features from Chicken Dinner Notebook | 0.11921477140699023, 0.11922039296180897
**Random Forest** |
All with Ranking | 0.0004156307577606391
Features from Chicken Dinner Notebook | 0.09820843398398921

(TODO) <br>
Because how dynamic a winPlacePerc value can be, custom prediction function needs to be developed

In [ ]:
# final_df[['Id', 'winPlacePerc']]